### PR TITLE
PM-11464 Add onboarding status to user Account to allow for root navigation to onboarding flow.

### PR DESCRIPTION
**Pull request by dseverns-livefront**
## 🎟️ Tracking
https://bitwarden.atlassian.net/browse/PM-11464

## 📔 Objective
- After a user has created an account or logged in to an existing account for the first time on the device (assuming uncleared storage) if the `OnboardingFlow` feature flag is on, set an onboarding status for the user
- Observe the user's state, if they have an Onboarding status that needs to be completed set the root nav state to reflect that
- Add navigation graph for onboarding flow

## 📸 Screenshots
*Control (onboarding off)*

https://github.com/user-attachments/assets/a0a16347-0813-4276-b42e-7a16e7856476

*onboarding on when account created*

https://github.com/user-attachments/assets/d6ff6eb4-b819-4732-8348-2ce196044e3f

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 1]
```diff
@@ -1559,6 +1571,16 @@ class AuthRepositoryImpl(
             ),
         )
         settingsRepository.hasUserLoggedInOrCreatedAccount = true
+
+        val shouldSetOnboardingStatus = featureFlagManager.getFeatureFlag(FlagKey.OnboardingFlow) &&
+            !settingsRepository.getUserHasLoggedInValue(userId = userId)
+        if (shouldSetOnboardingStatus) {
+            setOnboardingStatus(
+                userId = userId,
+                status = OnboardingStatus.NOT_STARTED,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1580]: Is this needed? Would we assume `NOT_STARTED` for null?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt [hunk 2]
```diff
@@ -113,6 +116,7 @@ fun RootNavScreen(
         is RootNavState.VaultUnlockedForFido2Save,
         is RootNavState.VaultUnlockedForFido2Assertion,
         is RootNavState.VaultUnlockedForFido2GetCredentials,
+        is RootNavState.OnboardingStep,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt:119]: This seems wrong, Shouldn't this be the ONBOARDING_GRAPH?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt [hunk 3]
```diff
@@ -60,6 +61,9 @@ class RootNavViewModel @Inject constructor(
         action: RootNavAction.Internal.UserStateUpdateReceive,
     ) {
         val userState = action.userState
+
+        val currentUserOnboardingStatus =
+            userState?.activeAccount?.onboardStatus ?: OnboardingStatus.COMPLETE
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt:66]: Ohh, you assume COMPLETE when null?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt (hunk 2 above)
Comment (brian-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt:119]: I'm agree, and I'm also curious about the naming...`OnboardingStep` sounds like a single step in the broader onboarding flow. Is there a reason not to just go with `Onboarding` for example?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt (hunk 3 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt:66]: Yes, thought process here is if it is null they created an account/logged in prior to this being in the app. So grandfathered in.

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt (hunk 2 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt:122]: I am actually now thinking that onboarding flow nav can just be all state based of this value and therefore should be flattened to a nav state per "step". Thank y'all!

Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavScreen.kt:122]: Just to add context from chat in standup: desired behavior is that each onboarding step would be a top level root destination, i.e. user does system back and exits the app. So will need to have this be one state per onboard status and navigate to that screen as the root.

**Comment by dseverns-livefront** [comment]
with new changes 

https://github.com/user-attachments/assets/cf1151a0-e72d-46f1-a27d-206c23934e13

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt (hunk 1 above)
Comment (david-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1579]: I see why we need to do this, carry on

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/UserStateJsonExtensions.kt [hunk 4]
```diff
@@ -171,6 +173,7 @@ fun UserStateJson.toUserState(
                     isUsingKeyConnector = userIsUsingKeyConnectorList
                         .find { it.userId == userId }
                         ?.isUsingKeyConnector == true,
+                    onboardStatus = onboardingStatusProvider(userId),
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/UserStateJsonExtensions.kt:176]: Could we make this nonnull and set the default here?

```kotlin
 onboardStatus = onboardingStatusProvider(userId) ?: OnboardingStatus.COMPLETE,
```

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/onboarding/OnboardingNavigation.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import androidx.navigation.NavController
+import androidx.navigation.NavGraphBuilder
+import androidx.navigation.NavHostController
+import androidx.navigation.NavOptions
+import androidx.navigation.navigation
+import com.x8bit.bitwarden.ui.auth.feature.accountsetup.SETUP_UNLOCK_ROUTE
+import com.x8bit.bitwarden.ui.auth.feature.accountsetup.setupUnlockDestination
+
+const val ONBOARDING_GRAPH_ROUTE: String = "onboarding_graph"
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/onboarding/OnboardingNavigation.kt:11]: Wanna tos a kdoc on here

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt (hunk 3 above)
Comment (david-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt:66]: Makes sense!

File: app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/OnboardingStatus.kt [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+     * The user is completing the account lock setup.
+     */
+    @SerialName("accountLockSetup")
+    ACCOUNT_LOCK_SETUP,
+
+    /**
+     * The user is completing the auto fill service setup.
+     */
+    @SerialName("autoFillSetup")
+    AUTO_FILL_SETUP,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/OnboardingStatus.kt:28]: autofill is one word

Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/model/OnboardingStatus.kt:28]: had a crisis about this, went this way cause "auto-fill." And there are areas where its "autofill" so idk. 😵

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/onboarding/OnboardingNavigation.kt (hunk 5 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/onboarding/OnboardingNavigation.kt:11]: I can, I don't see that for the other public route constants?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/UserStateJsonExtensions.kt [hunk 7]
```diff
@@ -171,6 +173,9 @@ fun UserStateJson.toUserState(
                     isUsingKeyConnector = userIsUsingKeyConnectorList
                         .find { it.userId == userId }
                         ?.isUsingKeyConnector == true,
+                    // If the user exists with no onboarding status we can assume they have been
+                    // using the app prior to the release of the onboarding flow.
+                    onboardingStatus = onboardingStatus ?: OnboardingStatus.COMPLETE,
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/UserStateJsonExtensions.kt:178]: 👍

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt [hunk 8]
```diff
@@ -1406,6 +1417,10 @@ class AuthRepositoryImpl(
             }
         }
 
+    private fun getOnboardingStatus(
+        userId: String,
+    ): OnboardingStatus? = authDiskSource.getOnboardingStatus(userId = userId)
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1422]: Is this ever used now?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt (hunk 1 above)
Comment (david-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1583]: I think it is but is this the correct spot to set this for things like TDE that defer final login?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/AuthDiskSourceExtensions.kt [hunk 9]
```diff
@@ -167,3 +169,24 @@ val AuthDiskSource.activeUserIdChangesFlow: Flow<String?>
...
         .map { it?.activeUserId }
         .distinctUntilChanged()
+
+/**
+ * Returns a [Flow] that emits every time the active user's onboarding status is changed
+ */
+@OptIn(ExperimentalCoroutinesApi::class)
+val AuthDiskSource.onboardingStatusChangesFlow: Flow<OnboardingStatus?>
+    get() {
+        return activeUserIdChangesFlow
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/AuthDiskSourceExtensions.kt:179]: Minor but can this just be an expression?

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/AuthDiskSourceExtensions.kt [hunk 10]
```diff
@@ -167,3 +169,24 @@ val AuthDiskSource.activeUserIdChangesFlow: Flow<String?>
...
+                activeUserId
+                    ?.let { this.getOnboardingStatusFlow(userId = it) }
+                    ?: flowOf(null)
+            }
+            .distinctUntilChanged()
+    }
+
+val AuthDiskSource.currentOnboardingStatus: OnboardingStatus?
+    get() = this
+        .userState
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/data/auth/repository/util/AuthDiskSourceExtensions.kt:190]: Can we add a test for this method

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/accountsetup/SetupAutoFillViewModel.kt [hunk 11]
```diff
@@ -72,11 +72,12 @@ class SetupAutoFillViewModel @Inject constructor(
     }
 
     private fun handleTurnOnLaterConfirmClick() {
-        sendEvent(SetupAutoFillEvent.NavigateToCompleteSetup)
+        // TODO PM-10631 record user chose to turn on later for settings badging.
+        // TODO PM-10632 update status to complete setup step.
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/accountsetup/SetupAutoFillViewModel.kt:76]: So we ditched the event and now we just need to update the state and then state-based nav just does the navigation for us?

File: app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt [hunk 12]
```diff
@@ -83,12 +84,22 @@ class RootNavViewModel @Inject constructor(
                     RootNavState.Auth
                 }
             }
-
```
Comment (david-livefront) [inline app/src/main/java/com/x8bit/bitwarden/ui/platform/feature/rootnav/RootNavViewModel.kt:86]: I think the formatter want this newline to be here

File: app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/accountsetup/SetupAutoFillViewModel.kt (hunk 11 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/ui/auth/feature/accountsetup/SetupAutoFillViewModel.kt:76]: yes! We will need to add another onboarding status for "FINISH_SETUP" for navigating to the final confirm screen, but yeah none of the onboarding will need event navigation. When we complete the user state will update then the root nav state will be "VaultUnlocked"

File: app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt (hunk 8 above)
Comment (dseverns-livefront) [reply app/src/main/java/com/x8bit/bitwarden/data/auth/repository/AuthRepositoryImpl.kt:1422]: ope ty!

**Review by david-livefront — APPROVED** [review]

**Pull request merged**